### ISSUE TITLE
T&T V3 - Dependency on DCSA_DOMAIN swagger

### ISSUE THREAD

**Issue by RababHayek** [dcsaorg/DCSA-OpenAPI#48]
Hello,

I'm Rabab Hayek, working as product owner in CMA CGM API Team.
I'm contributing to the implementation of T&T and OVS APIs and would like to share our feedback on the use of external swaggers such as DCSA_DOMAIN.

We can see totally the advantage of centralizing common object definitions.
However, if we put direct dependency on them within the swaggers to be implemented (such as T&T and OVS), this would lead to complex versioning management due to external evolving references...

In other words, shall these common definitions remain 'internal' to DCSA and have a monolithic T&T and OVS published swaggers ? Having independent swaggers will facilitate the carrier API maintenance...
 
I take advantage to check if the current T&T v3 is a stable version. We can see some external references which are not available, such as https://api.swaggerhub.com/domains/dcsaorg/DCSA_NOTIFICATION_DOMAIN/1.0.0#/components/schemas/notifications.

Thanks in advance

Best Regards,
Rabab

**Comment by HenrikHL** [comment]
Hi @RababHayek,
Welcome to this repository and great to hear CMA opinion :-)
We have had an internal talk about your concerns regarding dependencies. We don't like the idea to copy/paste - that is why we started using the Domains. I can see that you also favor centralizing of object definitions so let me start by sharing our thoughts:
The idea is to "lock" DCSA_DOMAIN version 1.0.0 as soon as EBL v1.0.0, T&T v2.1.0 or OVS v1.1.1 gets finalized. If any changes are required to the DCSA_DOMAIN after this time a new version (could be 1.0.1 depending on level of change) would be released. Whenever EBL, T&T or OVS are updated they would refer to the newest version of DCSA_DOMAIN.
At some point in time the DCSA_DOMAIN will require a major change (aka v2.0.0). API's using DCSA_DOMAIN don't necessarily need to change major version even though they refer to a new major DCSA_DOMAIN as long as the fields being referenced from the API's don't change. This would of course have to be managed.
It would be great if you could come up with some examples for your concern.
Next version of T&T will most likely be released in a version 2.1 ([https://app.swaggerhub.com/apis/dcsaorg/DCSA_TNT/2.1.0](https://app.swaggerhub.com/apis/dcsaorg/DCSA_TNT/2.1.0)) but this has not been finalized yet. Version 3 (which you mention) was an internal version and was not backward compatible. It was a suggestion to change event -> notification. It has been decided not to go forward with this approach. The "Notification" domain (which you link to) will also be discontinued.
DCSA is currently discussing with their members the changes proposed to version 2.1. We are aiming to make this version backward compatible with version 2.0 deprecating some fields and adding some new fields.

**Comment by RababHayek** [comment]
Hi @HenrikHL 

First of all, thanks for the clarification for T&T 2.1 version. Very good new (To be honest, I prefer event to notification 😊)
Concerning the dependency on DCSA_DOMAIN swagger, let me share the following example to be sure that I’m getting well your point:

**Initially, we have the below :**

- T&T v2.1.0 -> DCSA_DOMAIN 1.0.0
- OVS v1.1.1 -> DCSA_DOMAIN 1.0.0
- EBL  v1.0.0  -> DCSA_DOMAIN 1.0.0

**DCSA_DOMAIN 1.0.0 -> 1.0.1, then we might have the below situation**

- T&T v2.1.0 -> DCSA_DOMAIN 1.0.0
- OVS v1.1.1 -> DCSA_DOMAIN 1.0.1
- EBL  v1.0.0  -> DCSA_DOMAIN 1.0.1

**DCSA_DOMAIN 1.0.1 -> 2.0.0, then we might have the below situation**

- T&T v2.1.0 -> DCSA_DOMAIN 1.0.0
- OVS v1.1.1 -> DCSA_DOMAIN 1.0.1
- EBL  v1.0.0  -> DCSA_DOMAIN 2.0.0

=> Here, if we don’t force operational APIs (e.g. T&T, OVS & EBL) to be updated whenever a new DCSA_DOMAIN version is published, we can see that versioning management might become complex & somehow confusing for our API clients.

What do you think ?
Thanks in advance for your feedback :)

**Comment by HenrikHL** [comment]
Hi again @RababHayek,

Sorry for this late reply - I have been taking some days off during easter...

When DCSA_DOMAIN switches from 1.0.0 -> 1.0.1 then the situation will look like this:
- T&T v2.1.0 -> DCSA_DOMAIN 1.0.0
- OVS v1.1.**2** -> DCSA_DOMAIN 1.0.1
- EBL v1.0.**1** -> DCSA_DOMAIN 1.0.1

Notice that OVS and EBL need to be updated to a new version number in order to point to DCSA_DOMAIN 1.0.1 (I have updated OVS to 1.1.2 and EBL to 1.0.1 from your example above). As soon as an API or Domain is **published** - the API is **readonly**. The only way it can be modified is by increasing the version number and make the modifications.

When DCSA_DOMAIN changes from 1.0.1 -> 2.0.0 then it might look like this:
- T&T v2.1.0 -> DCSA_DOMAIN 1.0.0
- OVS v1.1.2 -> DCSA_DOMAIN 1.0.1
- EBL v1.0.2 -> DCSA_DOMAIN 2.0.0

If EBL is only pointing to new objects/components in DCSA_DOMAIN and is **NOT** utilizing any objects/components that have changed and are not backward compatible - then it is possible to release EBL in a version 1.0.2
If EBL is utilizing an object/component in DCSA_DOMAIN that has changed and is **NOT** backward compatible any more - then EBL needs to be released in a version 2.0.0 (to indicate that it is **NOT** backward compatible)
